skyportal/handlers/api/query.py

```diff
@@ -0,0 +1,264 @@
+import os
+import copy
+import yaml
+
+from langchain.embeddings.openai import OpenAIEmbeddings
+from langchain.vectorstores import Pinecone
+from typing import List, Optional
+import pinecone
+
+from baselayer.app.access import auth_or_token
+from baselayer.app.env import load_env
+from baselayer.log import make_log
+from ..base import BaseHandler
+from ...models import (
+    User,
+)
+
+_, cfg = load_env()
+log = make_log('query')
+
+
+# add in this new search method to the Pinecone class
+def search_sources(
+    self,
+    query: str,
+    k: int = 4,
+    filter: Optional[dict] = None,
+    namespace: Optional[str] = None,
+) -> List[dict]:
+    """Return pinecone documents most similar to query, along with scores.
+
+    Args:
+        query: Text to look up documents similar to.
+        k: Number of Documents to return. Defaults to 4.
+        filter: Dictionary of argument(s) to filter on metadata
+        namespace: Namespace to search in. Default will search in '' namespace.
+    Returns:
+        List of source dictionaries most similar to the query and score for each
+    """
+    if namespace is None:
+        namespace = self._namespace
+    query_obj = self._embedding_function(query)
+    sources = []
+    results = self._index.query(
+        [query_obj],
+        top_k=k,
+        include_metadata=True,
+        namespace=namespace,
+        filter=filter,
+    )
+    for res in results["matches"]:
+        try:
+            source = {
+                "id": res["id"],
+                "score": res["score"],
+                "metadata": res["metadata"],
+            }
+            sources.append(source)
+        except Exception as e:
+            log(f"Error: {e}")
+            continue
+    return sources
+
+
+setattr(Pinecone, 'search_sources', search_sources)
+
+# Preamble: get the embeddings and summary parameters ready
+# for now, we only support pinecone embeddings
+summarize_embedding_config = cfg[
+    'analysis_services.openai_analysis_service.embeddings_store.summary'
+]
+USE_PINECONE = False
+if (
```

Comment:
Maybe we log if something is missing to help admins?

Reply:
Done.